p2p/src/utils/requester.rs

```diff
@@ -0,0 +1,348 @@
+        let (participant, start, deadline) = self.requests.remove(&id)?;
+        let ids = self
+            .deadlines
+            .get_mut(&deadline)
+            .expect("deadline not found");
+        ids.remove(&id);
+        if ids.is_empty() {
+            self.deadlines.remove(&deadline);
+        }
+        Some((participant, start))
+    }
+
+    /// Ask for a participant to handle a request.
+    pub fn request(&mut self) -> Option<(PublicKey, ID)> {
+        // Look for a participant that can handle request
+        for (participant, _) in self.participants.iter() {
+            // Check if me
+            if *participant == self.crypto.public_key() {
+                continue;
+            }
+
+            // Check if excluded
+            if self.excluded.contains(participant) {
+                continue;
+            }
+
+            // Check if rate limit is exceeded (and update rate limiter if not)
+            if self.rate_limiter.check_key(participant).is_err() {
+                continue;
+            }
+
+            // Compute ID
+            //
+            // As long as we don't have u64 requests outstanding, this is ok.
+            let id = self.id;
+            self.id = self.id.wrapping_add(1);
+
+            // Record request issuance time
+            let now = self.runtime.current();
+            let deadline = now.checked_add(self.timeout).expect("time overflowed");
+            self.requests
+                .insert(id, (participant.clone(), now, deadline));
+            self.deadlines.entry(deadline).or_default().insert(id);
+            return Some((participant.clone(), id));
+        }
+        None
+    }
+
+    /// Resolve an outstanding request.
+    pub fn resolved(&mut self, id: ID) {
+        // Remove request
+        let Some((participant, start)) = self.remove(id) else {
+            return;
+        };
+
+        // Get elapsed time
+        let Ok(elapsed) = self.runtime.current().duration_since(start) else {
+            return;
+        };
+        let elapsed = elapsed.as_millis();
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past / 2 + elapsed / 2;
+        self.participants.put(participant, performance);
+    }
+
+    /// Timeout an outstanding request.
+    pub fn timeout(&mut self, id: ID) {
+        let Some((participant, _)) = self.remove(id) else {
+            return;
+        };
+
+        // Calculate new performance using exponential moving average
+        let Some(past) = self.participants.get(&participant) else {
+            return;
+        };
+        let performance = past / 2 + self.timeout.as_millis() / 2;
+        self.participants.put(participant.clone(), performance);
+    }
+
+    /// Cancel an outstanding request.
+    pub fn cancel(&mut self, id: ID) {
+        self.remove(id);
+    }
```

Comment:
Why is both `cancel` and `remove` needed? Am I missing a trait that needs to be impl'ed?

Suggest removing (no pun intended) `cancel` if it's the same except with no return value

Reply:
`remove` is a helper to keep everything up to date internally whereas `cancel` is meant to be publicly accessible. Fair point around why have both...